### PR TITLE
add Foreman host provider and client

### PR DESCRIPTION
**Pull request by BrandonArp**

### CODE REVIEW

File: app/com/arpnetworking/metrics/portal/hosts/impl/ForemanClient.java [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+         * @return This {@link Builder}.
+         */
+        public Builder setBaseUrl(final String value) {
+            _baseUrl = value;
+            return this;
+        }
+
+        @NotNull
+        @NotEmpty
+        private String _baseUrl;
```
Comment (vjkoskela) [inline app/com/arpnetworking/metrics/portal/hosts/impl/ForemanClient.java]: Should this be of type URI? It is supported by Jackson.

File: app/com/arpnetworking/metrics/portal/hosts/impl/ForemanClient.java [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+
+        private HostPageResponse(final Builder builder) {
+            _total = builder._total;
+            _subtotal = builder._subtotal;
+            _page = builder._page;
+            _perPage = builder._perPage;
+            _results = builder._results;
+        }
+
+        private final Integer _total;
```
Comment (vjkoskela) [inline app/com/arpnetworking/metrics/portal/hosts/impl/ForemanClient.java]: These should either be int (if not nullable) or Optional<Integer> if nullable.

File: app/com/arpnetworking/metrics/portal/hosts/impl/ForemanClient.java [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+             *
+             * @param value List of hosts
+             * @return this {@link Builder}
+             */
+            public Builder setResults(final List<ForemanHost> value) {
+                _results = value;
+                return this;
+            }
+
+            @JsonProperty("total")
```
Comment (vjkoskela) [inline app/com/arpnetworking/metrics/portal/hosts/impl/ForemanClient.java]: Not understanding the need for this annotation. Jackson should be using the setter which is ```setTotal``` which should map to the json key ```total``` so it seems redundant.

File: app/com/arpnetworking/metrics/portal/hosts/impl/ForemanClient.java [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+                return this;
+            }
+
+            /**
+             * Sets the list of hosts.  Required. Cannot be null.
+             *
+             * @param value List of hosts
+             * @return this {@link Builder}
+             */
+            public Builder setResults(final List<ForemanHost> value) {
```
Comment (vjkoskela) [inline app/com/arpnetworking/metrics/portal/hosts/impl/ForemanClient.java]: I've been moving towards Guava's ImmutableList and ImmutableSet more with much nicer semantics wrt copies. I would suggest either that or then a copy in the pojo constructor above.

File: app/com/arpnetworking/metrics/portal/hosts/impl/ForemanClient.java [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+             *
+             * @param value The name of the host
+             * @return this {@link Builder}
+             */
+            public Builder setName(final String value) {
+                _name = value;
+                return this;
+            }
+
+            @JsonProperty("name")
```
Comment (vjkoskela) [inline app/com/arpnetworking/metrics/portal/hosts/impl/ForemanClient.java]: Same question here regarding the annotation.

File: app/com/arpnetworking/metrics/portal/hosts/impl/ForemanHostProvider.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+import models.internal.Host;
+import models.internal.MetricsSoftwareState;
+import models.internal.impl.DefaultHost;
+import play.Configuration;
+
+import java.util.List;
+import javax.inject.Inject;
+
+/**
+ * @author Brandon Arp (brandon dot arp at smartsheet dot com)
```
Comment (vjkoskela) [inline app/com/arpnetworking/metrics/portal/hosts/impl/ForemanHostProvider.java]: Missing class comment.

File: app/com/arpnetworking/metrics/portal/hosts/impl/ForemanHostProvider.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+ */
+public class ForemanHostProvider extends UntypedActor {
```
Comment (vjkoskela) [inline app/com/arpnetworking/metrics/portal/hosts/impl/ForemanHostProvider.java]: Missing final. At least I hope we are not creating actor class trees ;)

File: app/com/arpnetworking/metrics/portal/hosts/impl/ForemanHostProvider.java [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+     * @param configuration Play configuration.
+     */
+    @Inject
+    public ForemanHostProvider(final HostRepository hostRepository, final Configuration configuration) {
+        _hostRepository = hostRepository;
+        getContext().system().scheduler().schedule(
+                ConfigurationHelper.getFiniteDuration(configuration, "hostProvider.initialDelay"),
+                ConfigurationHelper.getFiniteDuration(configuration, "hostProvider.interval"),
+                getSelf(),
+                "tick",
```
Comment (vjkoskela) [inline app/com/arpnetworking/metrics/portal/hosts/impl/ForemanHostProvider.java]: private static final String

File: app/com/arpnetworking/metrics/portal/hosts/impl/ForemanHostProvider.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+                getContext().dispatcher(),
+                getSelf());
+        _client = new ForemanClient.Builder()
+                .setBaseUrl(configuration.getString("hostProvider.foreman.baseUrl"))
+                .build();
+    }
+    @Override
```
Comment (vjkoskela) [inline app/com/arpnetworking/metrics/portal/hosts/impl/ForemanHostProvider.java]: Missing blank line above.

File: app/com/arpnetworking/metrics/portal/hosts/impl/ForemanHostProvider.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    public void onReceive(final Object message) throws Exception {
+        if ("tick".equals(message)) {
+            LOGGER.trace()
+                    .setMessage("Searching for added/updated hosts")
+                    .addData("actor", self())
+                    .log();
+            LOGGER.debug()
```
Comment (vjkoskela) [inline app/com/arpnetworking/metrics/portal/hosts/impl/ForemanHostProvider.java]: This double log seems weird. Should the other be in the host provider itself?

File: app/com/arpnetworking/metrics/portal/hosts/impl/ForemanHostProvider.java [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+                    .setMessage("Calling Foreman API to get hosts")
+                    .log();
+            Patterns.pipe(_client.getHostPage(1).wrapped(), context().dispatcher()).to(self(), self());
+        } else if (message instanceof ForemanClient.HostPageResponse) {
+            final ForemanClient.HostPageResponse response = (ForemanClient.HostPageResponse) message;
+            final List<ForemanClient.ForemanHost> results = response.getResults();
+            for (ForemanClient.ForemanHost host : results) {
```
Comment (vjkoskela) [inline app/com/arpnetworking/metrics/portal/hosts/impl/ForemanHostProvider.java]: host should be final

File: app/com/arpnetworking/metrics/portal/hosts/impl/ForemanClient.java (hunk 1 above)
Comment (BrandonArp) [inline app/com/arpnetworking/metrics/portal/hosts/impl/ForemanClient.java]: Sure.

File: app/com/arpnetworking/metrics/portal/hosts/impl/ForemanClient.java (hunk 2 above)
Comment (BrandonArp) [inline app/com/arpnetworking/metrics/portal/hosts/impl/ForemanClient.java]: Done

File: app/com/arpnetworking/metrics/portal/hosts/impl/ForemanClient.java (hunk 3 above)
Comment (BrandonArp) [inline app/com/arpnetworking/metrics/portal/hosts/impl/ForemanClient.java]: You're right.  This was from when I was being lazy and didn't have a builder.

File: app/com/arpnetworking/metrics/portal/hosts/impl/ForemanClient.java (hunk 4 above)
Comment (BrandonArp) [inline app/com/arpnetworking/metrics/portal/hosts/impl/ForemanClient.java]: Sure, done.

File: app/com/arpnetworking/metrics/portal/hosts/impl/ForemanClient.java (hunk 5 above)
Comment (BrandonArp) [inline app/com/arpnetworking/metrics/portal/hosts/impl/ForemanClient.java]: Fixed

File: app/com/arpnetworking/metrics/portal/hosts/impl/ForemanHostProvider.java (hunk 6 above)
Comment (BrandonArp) [inline app/com/arpnetworking/metrics/portal/hosts/impl/ForemanHostProvider.java]: Fixed

File: app/com/arpnetworking/metrics/portal/hosts/impl/ForemanHostProvider.java (hunk 7 above)
Comment (BrandonArp) [inline app/com/arpnetworking/metrics/portal/hosts/impl/ForemanHostProvider.java]: Fixed.

File: app/com/arpnetworking/metrics/portal/hosts/impl/ForemanHostProvider.java (hunk 9 above)
Comment (BrandonArp) [inline app/com/arpnetworking/metrics/portal/hosts/impl/ForemanHostProvider.java]: Fixed

File: app/com/arpnetworking/metrics/portal/hosts/impl/ForemanHostProvider.java (hunk 10 above)
Comment (BrandonArp) [inline app/com/arpnetworking/metrics/portal/hosts/impl/ForemanHostProvider.java]: Removed the double log.

File: app/com/arpnetworking/metrics/portal/hosts/impl/ForemanClient.java [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+            private Integer _total;
+
+            @NotNull
+            private Integer _subtotal;
+
+            @NotNull
+            private Integer _page;
+
+            @NotNull
+            @JsonProperty("per_page")
```
Comment (vjkoskela) [inline app/com/arpnetworking/metrics/portal/hosts/impl/ForemanClient.java]: I think this should be on the setter.

**Pull request merged**